src/cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -1695,6 +1700,126 @@ def _openssh_public_key_bytes(self, key):
                 serialization._ssh_write_string(public_numbers.encode_point())
             )
 
+    def generate_dh_parameters(self, generator, key_size):
+        if key_size < 512:
+            raise ValueError("DH key_size must be at least 512 bits")
+
+        dh_param_cdata = self._lib.DH_new()
+        self.openssl_assert(dh_param_cdata != self._ffi.NULL)
+        dh_param_cdata = self._ffi.gc(dh_param_cdata, self._lib.DH_free)
+
+        res = self._lib.DH_generate_parameters_ex(
+            dh_param_cdata,
+            key_size,
+            generator,
+            self._ffi.NULL
+        )
+        self.openssl_assert(res == 1)
+
+        return _DHParameters(self, dh_param_cdata)
+
+    def generate_dh_private_key(self, parameters):
+        dh_key_cdata = self._lib.DH_new()
+        self.openssl_assert(dh_key_cdata != self._ffi.NULL)
+        dh_key_cdata = self._ffi.gc(dh_key_cdata, self._lib.DH_free)
+
+        p = self._ffi.new("BIGNUM **")
```

Comment:
I think we can simplify this significantly if we bind `DHparams_dup` (that should be a separate PR though please)